src/components/tl/mlx5/mcast/tl_mlx5_mcast.h

```diff
@@ -384,6 +458,55 @@ static inline struct pp_packet* ucc_tl_mlx5_mcast_buf_get_free(ucc_tl_mlx5_mcast
     return pp;
 }
 
+static inline ucc_status_t ucc_tl_mlx5_mcast_post_user_recv_buffers(ucc_tl_mlx5_mcast_coll_comm_t *comm,
+                                                                    ucc_tl_mlx5_mcast_coll_req_t  *req,
+                                                                    int group_id, ucc_rank_t root,
+                                                                    int coll_type,
+                                                                    int count,
+                                                                    size_t offset)
+{
+    struct ibv_recv_wr *bad_wr = NULL;
+    struct ibv_recv_wr *rwr    = comm->call_rwr;
+    struct ibv_sge     *sge    = comm->call_rsgs;
+    struct pp_packet   *pp     = NULL;
+    uint32_t            i;
+
+    for (i = 0; i < count; i++) {
+        if (NULL == (pp = ucc_tl_mlx5_mcast_buf_get_free(comm))) {
+            tl_error(comm->lib, "not enought free pp packets to cover the entire message");
+            return UCC_ERR_NO_RESOURCE;
+        }
+
+        assert(offset % comm->max_per_packet == 0);
+        pp->packet_counter = offset / comm->max_per_packet;
+        pp->qp_id          = group_id;
+        rwr[i].wr_id       = ((uint64_t) pp);
+        rwr[i].next        = &rwr[i+1];
```

Comment:
nit:
```suggestion
        if (i < count -1) {
           rwr[i].next = &rwr[i+1];
        }
```
So its more readable, and we avoid `rwr[i-1].next = NULL;` and temporarily writing an illegal address

Reply:
The above suggested "if" branching should actually be fused with the else branch of the "`if (i == count - 1) {`" below

Reply:
i made the changes but setting to NULL is still required, as it might have junk value.

Reply:
> rwr[i].next = &rwr[i+

yes but it can be grouped with `sge[2*i + 1].length = req->last_pkt_len;`, or, better, you could use `count` instead of `i` for readability. I leave it up to you